회원가입-이메일인증

Jeonminje opened the pull request:
사용자 이메일로 메일 전송 및 전송코드 반환

Jeonminje committed:
SignUp-emailCheck

사용자 이메일로 메일 전송 및 전송코드 반환

Jeonminje committed:
SignUp-login

로그인 및 로그아웃을 세션을 이용해 구현

sunwootest commented on movieservice/src/main/java/study/movieservice/config/MailConfig.java:
```diff
@@ -0,0 +1,38 @@
+    public JavaMailSender javaMailService() {
+        JavaMailSenderImpl javaMailSender = new JavaMailSenderImpl();
+
+        javaMailSender.setHost("smtp.naver.com"); // 메인 도메인 서버 주소 => 정확히는 smtp 서버 주소
+        javaMailSender.setUsername("practice450"); // 네이버 아이디
+        javaMailSender.setPassword("alsdn120!"); // 네이버 비밀번호
```
민감한 정보는 application.properties에 저장하고 참조하시면 됩니다. properties 파일이 어떻게 동작하는지 알아보시면 좋을 것 같습니다. 민감 정보니까 빨리 처리해주세요.

Jeonminje replied:
처리하겠습니다!

sunwootest commented on movieservice/src/main/java/study/movieservice/config/MailConfig.java:
```diff
@@ -0,0 +1,38 @@
+        return javaMailSender;
+    }
+
+    private Properties getMailProperties() {
+        Properties properties = new Properties();
+        properties.setProperty("mail.transport.protocol", "smtp"); // 프로토콜 설정
+        properties.setProperty("mail.smtp.auth", "true"); // smtp 인증
+        properties.setProperty("mail.smtp.starttls.enable", "true"); // smtp strattles 사용
+        properties.setProperty("mail.debug", "true"); // 디버그 사용
+        properties.setProperty("mail.smtp.ssl.trust","smtp.naver.com"); // ssl 인증 서버는 smtp.naver.com
+        properties.setProperty("mail.smtp.ssl.enable","true"); // ssl 사용
+        return properties;
+    }
```
properties 정보를 굳이 클래스에 작성하실 필요는 없습니다. 참고로 JavaMailSender는 이미 빈으로 등록되어 있습니다.

Jeonminje replied:
수정하겠습니다!

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MailSender.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    // 인증 코드 생성.
+    public Integer createKey() {
+        Random random = new Random();
+        Integer randomNum = random.nextInt(888888) + 111111; // 111111 ~ 999999사이의 난수 생성을 한다.
+
+        return randomNum;
+    }
```
0으로 시작하는 코드는 없는건가요?

Jeonminje replied:
000000부터 시작하도록 수정하겠습니다

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MailSender.java:
```diff
@@ -0,0 +1,70 @@
+        message.addRecipients(MimeMessage.RecipientType.TO, email);// 보내는 대상
+        message.setSubject("Movie Systems 회원가입 이메일 인증입니다.");// 제목
+
+        String content =
+                "Movie Systems 회원가입 인증 메일입니다." +
+                        "<br><br>" +
+                        "인증 번호는 " + certificationNumber + "입니다." +
+                        "<br>" +
+                        "해당 인증번호를 인증번호 확인란에 기입하여 주세요.";
```
createMessage라는 메소드인데 회원가입 인증이라는 관심사에 의존되어 있는 것 같습니다.

sunwootest commented on movieservice/src/main/java/study/movieservice/service/MailSender.java:
```diff
@@ -0,0 +1,70 @@
+        try {
+            emailSender.send(message);
+        } catch (MailException e) {
+            e.printStackTrace();
+            throw new IllegalArgumentException();
```
이렇게 예외처리를 하면 사용자가 메일인증에 실패했다는 것을 어떻게 알 수 있죠?

Jeonminje replied:
수정하겠습니다!

Jeonminje replied:
인증 코드를 클라이언트쪽에 넘겨주어 서버에서 확인 할 필요없이 프론트에서 확인을 하는 방법은 안좋은 방법일까요?

sunwootest replied:
보안과 관련된 내용은 가능하면 서버측으로 숨기는 것이 좋습니다 :)

Jeonminje replied:
넵 감사합니다!

Jeonminje committed:
application.properties 파일 삭제

보안을 위해 삭제하였습니다.

Jeonminje committed:
Merge branch 'SignUp-save-minje' into SignUp-emailCheck-minje

# Conflicts:
#	build/classes/java/main/study/movieservice/config/MailConfig.class
#	build/classes/java/main/study/movieservice/service/MailSender.class
#	build/tmp/compileJava/compileTransaction/stash-dir/MemberController.class.uniqueId0
#	movieservice/build/classes/java/main/study/movieservice/controller/MemberController.class
#	movieservice/build/tmp/compileJava/previous-compilation-data.bin
#	movieservice/src/main/java/study/movieservice/controller/MemberController.java
#	src/main/java/study/movieservice/config/MailConfig.java
#	src/main/java/study/movieservice/service/MailSender.java

Jeonminje committed:
Merge remote-tracking branch 'origin/SignUp-emailCheck-minje' into SignUp-emailCheck-minje

Jeonminje committed:
이메일 인증코드 전송 수정 및 코드 확인 로직 추가

- 기존 JavaMailSender를 설정한 뒤 Bean으로 수동 등록 하던 것에서
properties 파일을 이용해 설정을 숨기고,
JavaMailSender는 스프링이 등록 해주는 것을 사용하는 방향으로 수정하였습니다.

- controller.GeneralExceptionHandler 클래스
이메일 전송 에러 처리 추가했습니다.

- controller.MemberController 클래스
이메일 전송 및 이메일 코드 확인 관련 매핑 추가했습니다.

- repository.RedisUtil 클래스 추가
이메일 코드를 보관하기 위해 redis를 사용하였으며, key는 이메일, value는 코드를 사용하고
유효시간을 설정하도록 하였습니다.
getData() 함수는 Optional을 반환하여 데이터가 없을 시,
호출한 MailService에서 null처리를 해주도록하였습니다.

- service.MailService 클래스
createMessage() 에서 메시지내용과 제목을 직접 작성하던 것을 따로 함수로 작성하였습니다.
필드에 있던 인증코드 변수를 동시성 문제를 고려해 함수안으로 넣었고,
함수간에 파라미터로 생성한 인증코드를 전달 해 사용하는 방식으로 변환했습니다.
checkEmailCode() 함수를 추가해 전송한 코드와 일치하는지 확인하도록 하였습니다.
추가로 111111~999999였던 코드 범위를 000000~999999 범위까지 늘렸습니다.

Jeonminje commented:
커밋 했습니다!
기존 회원가입 및 아이디 중복확인 approve 해주셔서 email-check 레포지토리에 merge하고 조언 해주신 내용 수정 하였습니다

Jeonminje committed:
잘못한 커밋 삭제

예전에 잘못 커밋한 파일입니다.

Jeonminje committed:
잘못한 커밋 삭제

잘못된 커밋입니다.

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -30,4 +33,27 @@ public ResponseEntity<String> loginIdCheck(@RequestParam String loginId) throws
+    @ResponseBody
+    @PostMapping("/email-check")
+    public ResponseEntity<String> memberEmailCheck(@RequestParam String email) {
+        try {
+            mailService.sendMail(email);
+        } catch (Exception e) {
+            throw new MailSendException("이메일 전송에 실패하였습니다.");
+        }
```
presentation layer에 비즈니스 로직의 예외처리를 위임할 필요가 있을까요? 이런 예외 처리는 Buisiness Layer내에서 처리하는 것이 나아보입니다.

Jeonminje replied:
넵 감사합니다!

sunwootest commented on src/main/java/study/movieservice/repository/RedisUtil.java:
```diff
@@ -0,0 +1,31 @@
+public class RedisUtil {
+
+    private final StringRedisTemplate redisTemplate;
+
+    public Optional<String> getData(String email){
+        ValueOperations<String,String > store = redisTemplate.opsForValue();
+        return Optional.ofNullable(store.get(email));
+    }
```
이 메소드와 dleteData 메소드는 굳이 Util클래스로 만들 필요 없이 바로 RedisTemplate을 호출하는 편이 나아보이네요. 그리고 Util 클래스면 static으로 선언하는 편이 낫지 않을까요?

sunwootest replied:
RedisUtil인데 이메일에 종속적인 메소드들만 있는 것 같습니다.

Jeonminje replied:
조언에 따라 수정해보도록 하겠습니다

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+        String content = messageContent(emailCode);
+        message.setText(content, "utf-8", "html");
+
+
+        message.setFrom(new InternetAddress("practice450@naver.com","MovieSystems"));
```
이런 정보는 프로퍼티 파일로 분리시키는게 나아보입니다. 송신자 정보가 바뀔 때마다 코드를 수정하는 일은 번거로울 수 있거든요.

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+
+    private final JavaMailSender emailSender;
+    private final RedisUtil redisUtil;
+
+    public MimeMessage createMessage(String email, String emailCode) throws MessagingException, UnsupportedEncodingException {
```
예외를 꼭 던질 필요가 있을까요? ExceptionHandler도 사용할줄 아시는데 예외전환전략을 사용해도 괜찮아 보입니다.

Jeonminje replied:
수정하겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+        return key.toString();
+    }
+
+    @Transactional
+    public void sendMail(String email) throws Exception {
```
해당 메소드에 Transactional을 왜 사용하셨나요? 추가적으로 Transactional 메소드를 사용할 때 주의점이 무엇이 있을까요?

Jeonminje replied:
주의점으로는 기본적으로 Checked Exception에 대해서는 롤백시키지 않도록 되어있다는 것과 트랜잭션 전파관련 문제가 있을 것 같습니다!

Jeonminje replied:
원래 의도는 redis에 이메일 인증코드를 저장하고, 메일을 보낼 때 이상이 생기면, 인증코드 저장로직을 rollback 해줘야 한다고 생각해서 Transactional을 사용하였습니다. 하지만 적용이 잘 되지않는 문제가 있었습니다. 주의점을 잘 파악해 적용시켜보겠습니다!

sunwootest replied:
메일을 보낼 때 문제가 생기면 예외처리를 해주면 될 것 같네요. redis의 트랜잭션을 보장하기 위한 방법이 몇가지가 있는데 시간나면 이런 것도 알아보시면 좋을 것 같습니다.

sunwootest commented on src/main/resources/application.properties:
```diff
@@ -1,10 +1,21 @@
+#MyBatis in xml convenient resultType.
 mybatis.type-aliases-package=study.movieservice.domain
 #db and java convert snake-case, camel-case
 mybatis.configuration.map-underscore-to-camel-case=true
```
해당 파일이 gitignore 적용이 안되네요 :)

Jeonminje replied:
수정하겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+        return message;
+    }
+
+    private String messageSubject() {
+        return "Movie Systems 회원가입 이메일 인증입니다.";
+    }
+
+    private String messageContent(String emailCode) {
+        return "Movie Systems 회원가입 인증 메일입니다." +
+                "<br><br>" +
+                "인증 번호는 " + emailCode + "입니다." +
+                "<br>" +
+                "해당 인증번호를 인증번호 확인란에 기입하여 주세요.";
+    }
```
sendMail이라는 메소드에서 회원가입이라는 관심사는 잘분리하셨습니다. 하지만 MailService라는 클래스 안에 여전히 회원가입이 의존하고 있네요.

Jeonminje replied:
수정하도록하겠습니다

Jeonminje committed:
잘못된 커밋 삭제

보안관련 문제로 삭제하겠습니다.

Jeonminje committed:
-MemberController
 MailService에서 예외처리로 변경

- RedisUtil
 삭제 후 MailService내에서 StringRedisTemplate 주입받아 사용

- MailService
 기존 발신자 정보를 프로퍼티파일로 분리했습니다.
 createMessage()에서 던지던 예외를 unchecked예외로 변환 후 던져 ExceptionHandler에서 처리했습니다.
 Transactional 어노테이션 삭제했습니다. (redis는 rollback을 지원하지 않고, 인증코드는 3분뒤에 사라지고 또한 사용자는 메시지 전송 실패를 전달 받아 인증코드 재발급 받을 시 overwrite되므로 상관없다 생각했습니다.)
 기존 MailService에서 작성하던 메시지 내용 및 제목을 MessageForm 클래스를 만들어 더이상 의존하지 않도록 처리했습니다.

Jeonminje commented:
커밋했습니다!

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+@RequiredArgsConstructor
+public class MailService {
+
+    private final JavaMailSender emailSender;
+    public final StringRedisTemplate redisTemplate;
+    @Value("${spring.mail.username}")
+    private String address;
+    @Value("${sender}")
+    private String sender;
```
모두 private final로 선언해도 될 것 같은데 이렇게 선언한 이유가 있으신가요?

Jeonminje replied:
정보를 properties파일에 숨기다보니 세부적인 것을 신경쓰지 못한것 같습니다 한번 더 살펴보겠습니다

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+        ValueOperations<String,String > store = redisTemplate.opsForValue();
+        String code = store.get(email);
+
+        if( code != null && code.equals(emailCode)){
```
조건문 안에 불필요한 공백이 있네요!

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,81 @@
+        return key.toString();
+    }
+
+    public void sendMail(String email) {
+        String emailCode = createKey();
+        MimeMessage message = createMessage(email, emailCode);
+        try{
+            ValueOperations<String, String> store = redisTemplate.opsForValue();
+            Duration duration = Duration.ofSeconds(60*3L);
+            store.set(email, emailCode, duration);
+
+            emailSender.send(message);
+        }catch (Exception e){
+            throw new MailSendException("이메일 전송에 실패하였습니다.");
+        }
+    }
```
여전히 회원가입이라는 관심사에 의존하고 있네요. 마지막으로 한번 더 고민하시고 메일 전송이라는 서비스와 회원가입이라는 서비스를 완전히 분리해보도록 하세요 :)

Jeonminje replied:
수정하겠습니다!

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -30,4 +32,23 @@ public ResponseEntity<String> loginIdCheck(@RequestParam String loginId) throws
     public void memberJoin(@RequestBody MemberDTO memberDTO) {
         memberService.addMember(memberDTO);
```
이것도 고민해볼 여지가 있는 것이라 알려드립니다. MemberDto에서 Dto를 꼭 붙여야할까요?

Jeonminje replied:
넵 고민해보겠습니다!

Jeonminje committed:
# 관심사 분리

## MemberController
- memberEmailCheck() 함수에서 mailService 클래스의 sendMail() 를 사용하던것에서 MemberService 클래스에 sendSignUpMessage()함수를 사용하여 회원가입 인증코드 전송.

## MailService
- address 와 sender 변수를 private final로 바꾸고 생성자의 파라미터에서 @value를 통해 properties파일에 있는 데이터를 주입해주었습니다.
- 기존 createMessage() 함수를 createSignUpMessage() 함수로 변경해 회원가입시 사용하는 메시지를 생성해주는 함수로 지정해주고, 메시지 생성에 문제가 없다면 storeKey() 를 통해 레디스에 이메일과 코드를 저장해주었습니다.
- 기존 sendMail() 함수가 회원가입 관심사에 의존적이다 판단해, MimeMessage 인스턴스를 생성해 넘겨주면 단순히 전송해주는 역할로 바꾸었습니다.

## MemberService
- 필드에 MailService를 주입받아 사용하는 것이 추가되었습니다.
- sendSignUpMessage() 함수를 추가하였습니다. MailService 클래스의 함수를 이용해 인증코드 생성, MimeMessage 객체생성 및 전송을 담당하는 역할 입니다.

## MessageForm
- 함수명을 messageSubject() -> signUpMessageSubject() , messageContent() -> signUpMessageContent() 로 책임에 따라 함수명을 지어줬습니다.

Jeonminje commented:
커밋했습니다!

Jeonminje committed:
Merge branch 'SignUp-emailCheck-minje' into SignUp-login-minje

# Conflicts:
#	movieservice/src/main/java/study/movieservice/service/MemberService.java
#	src/main/java/study/movieservice/controller/LoginController.java
#	src/main/java/study/movieservice/domain/SessionConst.java

Jeonminje committed:
로그인 기능 수정 및 이메일 중복확인 기능과 merge

## LoginController
- 로그인과 로그아웃 요청이 들어오면 MemberService의 함수로 처리하였습니다.

## MemberMapper
- getMember() 함수를 추가하였습니다.

## MemberService
- logIn() logOut() 함수를 추가해 세션을 이용한 로그인, 로그아웃을 구현하였습니다.

Jeonminje commented:
## MemberController
- memberEmailCheck() 함수에서 mailService 클래스의 sendMail() 를 사용하던것에서 MemberService 클래스에 sendSignUpMessage()함수를 사용하여 회원가입 인증코드 전송.

## MailService
- address 와 sender 변수를 private final로 바꾸고 생성자의 파라미터에서 https://github.com/value를 통해 properties파일에 있는 데이터를 주입해주었습니다.
- 기존 createMessage() 함수를 createSignUpMessage() 함수로 변경해 회원가입시 사용하는 메시지를 생성해주는 함수로 지정해주고, 메시지 생성에 문제가 없다면 storeKey() 를 통해 레디스에 이메일과 코드를 저장해주었습니다.
- 기존 sendMail() 함수가 회원가입 관심사에 의존적이다 판단해, MimeMessage 인스턴스를 생성해 넘겨주면 단순히 전송해주는 역할로 바꾸었습니다.

## MemberService
- 필드에 MailService를 주입받아 사용하는 것이 추가되었습니다.
- sendSignUpMessage() 함수를 추가하였습니다. MailService 클래스의 함수를 이용해 인증코드 생성, MimeMessage 객체생성 및 전송을 담당하는 역할 입니다.

## MessageForm
- 함수명을 messageSubject() -> signUpMessageSubject() , messageContent() -> signUpMessageContent() 로 책임에 따라 함수명을 지어줬습니다.

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,91 @@
+    private final String address;
+    private final String sender;
+
+    @Autowired
+    public MailService(JavaMailSender emailSender, StringRedisTemplate redisTemplate,
```
생성자가 하나일 경우 @Autowired를 생략해도 생성자 주입이 이루어집니다.

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,91 @@
+        store.set(email, emailCode, duration);
+    }
+
+    public String createKey() {
+        StringBuilder key = new StringBuilder("");
+        Random random = new Random();
+
+        for(int i=0;i<6;i++){
+            Integer randomNum = random.nextInt(9);
+            key.append(randomNum);
+        }
+
+        return key.toString();
+    }
```
이 부분도 분리하면 좋긴 하겠지만 토이프로젝트니까 일단은 놔두셔도 될 것 같습니다 :)

Jeonminje replied:
감사합니다!

sunwootest commented on src/main/java/study/movieservice/service/MessageForm.java:
```diff
@@ -0,0 +1,18 @@
+public final class MessageForm {
+
+    private MessageForm(){
+    }
+    public static String signUpMessageSubject() {
```
메소드 사이에 개행이 필요해보이네요.

Jeonminje replied:
수정하겠습니다!

sunwootest reviewed (commented):
관심사를 분리해야한다는 것을 email 구현을 통해서 어느정도 이해하신 것 같습니다. 아주 완벽하게 관심사를 분리할 순 없지만 토이프로젝트이기 때문에 이 정도 수준으로만 분리하면 충분해보이네요 :) 임시 approve 드립니다.

Commit:
지켜지지 않은 코드 컨벤션 수정

Commit:
Merge branch 'SignUp-emailCheck-minje' into SignUp-login-minje

Commit:
로그인 컨트롤러 및 로그인 서비스 관련 함수 수정

## LoginController
- HttpServletRequest에 의존하던 코드를 HttpSession을 직접 받아 사용하는 것으로 변경하였습니다.
- 로그인 실패시 200코드를 반환해주던 것에서 오류코드를 리턴해주도록 변경하였습니다.
- 아이디와 비밀번호를 받기위해 LoginParam 클래스를 추가해 DTO로 이용하였습니다.

## LoginParam
- 아이디와 비밀번호를 받기위한 DTO로 이번에 추가했습니다.

## MemberMapper.xml
- 테이블 명을 members -> member로 바꾸고 관련 쿼리를 수정하였습니다.

## SessionConst
- 세션의 키값을 가지고 있는 클래스를 만들어 주었습니다. (닉네임, 로그인아이디)

## MemberService
- logIn()에서 기존 세션에 key = 아이디, value = 닉네임 으로 저장하던 것에서 각각 value로 저장하도록 수정했습니다. 또한 기존 HttpServletRequest 관련 코드를 HttpSession으로 수정하였습니다.
- logOut() : 기존 loginId를 파라미터로 받아 로그인아이디가 key로 저장된 세션에서 삭제를 해주는 방식에서 HttpSession.invalidate() 함수를 이용해 로그아웃 시 세션을 비우는 방식으로 수정하였습니다.

Jeonminje committed:
예외 추가

## exception 폴더를 만들어 예외 추가 ( 모두 런타임 예외 )
- LoginIncorrectException : 아이디나 비밀번호가 일치하지 않을 경우 발생하는 예외 입니다.
- LoginSetAttributeException : 세션의 key, value를 세팅할 때 예외가 발생하면 이 예외로 변환을 해줍니다.
- LogoutException : 로그아웃시 발생하는 예외 입니다.

## LoginController
- 로그인이 실패하면 예외를 발생시켜 ExceptionHandler에서 응답을 하도록 해주었습니다.

## MemberMapper
- getMember() 함수의 명을 수정하였습니다.

## MemberService
- logOut(), logIn() 함수에서 HttpSession 관련 로직을 분리하였습니다. SessionManager를 빈으로 주입받아 사용 합니다.

## SessionManager
- HttpSession을 빈으로 주입받아 세션관련 로직을 수행해줍니다.

Jeonminje committed:
GeneralExceptionHandler 추가 및 예외 처리 수정
1. 기존 controller에서 ResponseEntity 를 리턴하던 것에서 String을 리턴하는 것으로 방식을 바꾸고 @ResponseStatus 어노테이션을 이용하여 상태를 응답해주었습니다.
2. controller에 있던 비즈니스 로직을 모두 제거하였습니다. ( 예외는 모두 service단에서 던져주었습니다.)
3. EexceptionMessageConst 이넘 클래스를 생성하여 예외 메시지를 분리해주었습니다.
4. 기존 RuntimeException을 상속하여 직접 만들어 주었던 예외들을 모두 삭제, 및 만들어져 있는 예외를 사용하는 방안으로 변환해주었습니다.
5. 로그인 시, 세션에 저장하는 값을 로그인아이디가 아닌 primary key 값인 memberId를 저장해주었습니다.

Jeonminje committed:
log-in시 session에 nickname과 memberId를 저장하던 것을 닉네임은 저장하지 않는 코드로 수정.

Jeonminje committed:
SessionConst의 nickname관련 상수 제거

Jeonminje committed:
Merge pull request #11 from Jeonminje/SignUp-login-minje

로그인 기능
